runtime-modules/bounty/src/lib.rs

```diff
@@ -773,138 +866,266 @@ decl_module! {
+            ensure!(
+                matches!(current_bounty_stage, BountyStage::Funding{..}),
+                Self::unexpected_bounty_stage_error(current_bounty_stage),
+            );
+
+            //contribution is adjusted to prevent exceeding target funding.
+            let (is_target_funding_reached,
+                is_first_contribution,
+                adjusted_amount,
+                adjusted_contribution) =
+                Self::get_adjusted_contribution(&bounty_id, &bounty, &funder, amount);
+            let state_bloat_bond_amount = adjusted_contribution.funder_state_bloat_bond_amount;
+
+            let transfer_amount = match is_first_contribution {
+                //If It's the first contribution, the transfer amount includes the state bloat bond amount.
+                true => adjusted_amount.saturating_add(state_bloat_bond_amount),
+                false => adjusted_amount
+            };
+
+            bounty_funder_manager.validate_balance_sufficiency(transfer_amount)?;
 
             //
             // == MUTATION SAFE ==
             //
 
-            Self::return_bounty_cherry_to_creator(bounty_id, &bounty)?;
+            bounty_funder_manager
+                .transfer_funds_to_bounty_account(bounty_id, transfer_amount);
 
-            Self::remove_bounty(&bounty_id);
+            let new_milestone = Self::get_bounty_milestone_on_funding(
+                is_target_funding_reached,
+                bounty.milestone);
 
-            Self::deposit_event(RawEvent::BountyCanceled(bounty_id, creator));
+            // Update bounty record.
+            <Bounties<T>>::mutate(bounty_id, |bounty| {
+                //Updates only the funds not the bloat bond.
+                bounty.total_funding = bounty.total_funding.saturating_add(adjusted_amount);
+                bounty.milestone = new_milestone;
+            });
+
+            // Update member funding record
+            <BountyContributions<T>>::insert(bounty_id, funder.clone(), adjusted_contribution);
+
+            // Fire events.
+            Self::deposit_event(RawEvent::BountyFunded(bounty_id, funder, adjusted_amount));
+            if is_target_funding_reached{
+                Self::deposit_event(RawEvent::BountyMaxFundingReached(bounty_id));
+            }
         }
 
-        /// Vetoes a bounty.
-        /// It returns a cherry to creator and removes bounty.
+        /// Terminates a bounty with perpetual working period.
+        /// It returns a oracle cherry to creator.
         /// # <weight>
         ///
         /// ## weight
         /// `O (1)`
         /// - db:
         ///    - `O(1)` doesn't depend on the state or parameters
         /// # </weight>
-        #[weight = WeightInfoBounty::<T>::veto_bounty()]
-        pub fn veto_bounty(origin, bounty_id: T::BountyId) {
+        #[weight = WeightInfoBounty::<T>::terminate_bounty()]
+        pub fn terminate_bounty(origin, bounty_id: T::BountyId) {
```

Comment:
NOTE: Recall my comment above about dropping `cancel_bounty`, and allowing creator to call `terminate_bounty` during funding stage (even after contributions).

Now to the main point:

# Bug

As discussed:

1. Bounty is in funding period.
2. Has no contributions.
3. Has no oracle reward.
4. Is terminated.

At this point, we will return cherry to creator, and remains in `BountyFailed` stage, and noone can do any final call for bounty to actually get removed.

# Fix

- If there are contributions, then just do as now: just go to `FailedBounty`.
- If instead there are no contributions, then
   - return cherry to oracle
   - if unclaimed oracle reward exists, then go to `FailedBounty`: this will allow withdrawal of reward and final cleanup
   - if there is no unclaimed oracle reward, then `remove_bounty`, which does cleanup